docs/_docs/reference/experimental/typeclasses.md

```diff
@@ -0,0 +1,776 @@
+
+**Example**
+
+```scala
+  trait Ord:
+    type Self
+
+  trait SemiGroup:
+    type Self
+    extension (x: Self) def combine(y: Self): Self
+
+  trait Monoid extends SemiGroup:
+    def unit: Self
+
+  trait Functor:
+    type Self[A]
+    extension [A](x: Self[A]) def map[B](f: A => B): Self[B]
+
+  trait Monad extends Functor:
+    def pure[A](x: A): Self[A]
+    extension [A](x: Self[A])
+      def flatMap[B](f: A => Self[B]): Self[B]
+      def map[B](f: A => B) = x.flatMap(f `andThen` pure)
+
+  def reduce[A: Monoid](xs: List[A]): A =
+    xs.foldLeft(Monoid.unit)(_ `combine` _)
+
+  trait ParserCombinator:
+    type Self
+    type Input
+    type Result
+    extension (self: Self)
+      def parse(input: Input): Option[Result] = ...
+
+  def combine[A: ParserCombinator, B: ParserCombinator { type Input = A.Input }] = ...
+```
+
+**Advantages**
+
+ - Avoids repetitive type parameters, concentrates on what's essential, namely the type class hierarchy.
+ - Gives a clear indication of traits intended as type classes. A trait is a type class
+   if it has type `Self` as a member
+ - Allows to create aggregate type classes that combine givens via intersection types.
+ - Allows to use refinements in context bounds (the `combine` example above would be very awkward to express using the old way of context bounds expanding to type constructors).
+
+`Self`-based context bounds are a better fit for a dependently typed language like Scala than parameter-based ones. The main reason is that we are dealing with proper types, not type constructors. Proper types can be parameterized, intersected, or refined. This makes `Self`-based designs inherently more compositional than parameterized ones.
+
+
+
+**Details**
+
+When a trait has both a type parameter and an abstract `Self` type, we
+   resolve a context bound to the `Self` type. This allows type classes
+   that carry type parameters, as in
+
+```scala
+trait Sequential[E]:
+  type Self
+```
+
+Here,
+```scala
+[S: Sequential[Int]]
+```
+should resolve to:
+```scala
+[S](using Sequential[Int] { type Self = S })
+```
+and not to:
+```scala
+[S](using Sequential[S])
+```
+
+**Discussion**
+
+ Why not use `This` for the self type? The name `This` suggests that it is the type of `this`. But this is not true for type class traits. `Self` is the name of the type implementing a distinguished _member type_ of the trait in a `given` definition. `Self` is an established term in both Rust and Swift with the meaning used here.
+
+ One possible objection to the `Self` based design is that it does not cover "multi-parameter" type classes. But neither do context bounds! "Multi-parameter" type classes in Scala are simply givens that can be synthesized with the standard mechanisms. Type classes in the strict sense abstract only over a single type, namely the implementation type of a trait.
+
+
+## Auxiliary Type Alias `is`
+
```

Comment:
So from [here](https://circe.github.io/circe/codecs/custom-codecs.html) you have `implicit val decodeFoo: Decoder[Thing] = new Decoder[Thing]{...}`  which would be
```scala
given Thing is Decoder:
  ...
```
and you would like
```scala
given Thing has Decoder:
  ...
```
We could have both as it is just a Predef type alias.

But we could also if `is` is too itchy:
```scala
type Decoded = Decoder
given Thing is Decoded:
  ...
```
and it reads a bit better.

Maybe it's because I'm a non-native English speaker that I'm more pragmatic about 
`given Thing is Decoder` as it kind of *is* a Decoder in the sense that Thing now can do decoding, which is what the Decoder is doing behind the scenes.

Reply:
> We could have both as it is just a Predef type alias.

Sure, but I do think `has` is a much better default than `is` if you look at popular type classes in Scala 2 today. Also it is IMO better to not give too much choice and stick with the most logical choice looking at how Scala is used.

> given Thing is Decoder

IMO if you make `Thing` concrete for actual types like `Int`, `String` and `Person` the `is` stands completely out of place. Now again, I know this is nitpicking. BUt I had the idea that there was much more consensus for `has` the last time this discussion was held and it is ignored (again).

Reply:
(Aside: this thread comments a commit that is part of early experimentation even before a subsequent improvement proposal is submitted. No comments will be "ignored" as we will continue to discuss all trade-offs. All previous discussions is input to the SIP committee as we come to decision-making about this so I don't think it is fair to say "ignored (again)" here; this thread is open to the whole planet earth to read and comment and the SIP committee's intention is to be benevolent of all factual input and try to make good trade-offs based on what is included both in written proposals and in pointers to open discussions etc.)

Anyways, back to the factual matters: The proposal regards new type classes that use the new Self-type-member scheme; the old style wont go away as type parameters and given-using wont go away. So the naming of type-param-based type classes can stay the same. For new type-member-based type classes the naming can take the `is`-type-alias into account. And the nice thing with `is` being just a type alias in Predef is that anyone can define `has` in a similar way locally.

I actually think this makes sense:
```scala
given Int is Decoder
```
if you think of it as the Int class is given the ability to be a Decoder of serialized integer values. You can read it as "given that Int is a Decoder then we can use Int to decode serialized Int values".

Reply:
> The proposal regards new type classes that use the new Self-type-member scheme; the old style wont go away as type parameters and given-using wont go away.

That will mean that Scala will have 2 different ways of defining and using typeclasses. And in reality you can add implicit to that mix as well for as long as it is still being supported.

Reply:
Well, that's a price to pay when embracing both innovation, stability and ease of migration. We should consider and balance trade-offs between stability and innovation in each case. And, after these experiments are turned into a proposal the SIP committee will take all input into account and vote to try to make the best decision. But this is still early days and I think it is good if these experiments that try to leverage the recent abilities of the type system can get out there in a reasonably stable "best-possible-for-now"-version for everyone to play with under a flag to give feedback and surface unanticipated problems etc.